iroha_p2p/src/peer.rs

```diff
@@ -247,18 +253,30 @@ where
 #[async_trait::async_trait]
 impl<T, K, E> Actor for Peer<T, K, E>
 where
-    T: Encode + Decode + Send + Clone + 'static,
+    T: Encode + Decode + Send + Sync + Clone + 'static,
     K: KeyExchangeScheme + Send + 'static,
     E: Encryptor + Send + 'static,
 {
     async fn on_start(&mut self, ctx: &mut Context<Self>) {
+        info!("Starting peer for {}", &self.id.address);
+        let mut count = 0;
         while self.state != State::Ready {
             if let Err(e) = self.handshake().await {
-                warn!(%e, "Error connecting to peer {}, bailing.", &self.id.address);
-                break;
+                warn!("Error connecting to peer {}. {}", &self.id.address, e);
+                if count < MAX_CONNECTION_RETRY_COUNT {
+                    count += 1;
+                    let delay = CONNECTION_RETRY_PERIOD * count;
+                    debug!(%e, "Error connecting to peer {}, waiting {}ms before retry.", &self.id.address, delay);
+                    time::sleep(Duration::from_millis(delay)).await;
+                    continue;
+                }
```

Comment:
I have a general question about implementation here. If the max connection retry count is exceeded and peer1 stops trying to connect to peer2, and after that peer2 becomes alive - will peer2 be able to initialize handshake with peer1 by itself?

Reply:
Yes, they generally connect all to each other. And we will need some mechanism to distinguish tween-connects and close them. This will be in one of next PR's.

Reply:
Can you create an issue and link it here for removing twin connects?

Reply:
Sure, here it is:
https://github.com/hyperledger/iroha/issues/1272